### PR TITLE
TIKA-4766: minimal typed Document parse contract for tika-grpc (stage 1)

### PR DESCRIPTION
**Pull request by krickert**
First slice of the typed Document contract — stage 1 of the sequence proposed by @nddipiazza in #2921: the `Document` envelope, Dublin Core `DocumentMetadata`, the lossless tagged metadata tail, and `ParseStatus`. Generic transformer only; no blocks/markdown, no embedded recursion, no `FormatCategory`; `ParseBytes` is deferred to its own JIRA.

**The reply is additive.** `FetchAndParseReply.fields`/`status` are untouched and `document` is a new field alongside — no client breaks (the "stage the break" approach from #2916). The legacy fields map keeps carrying the flat content until the content tree lands in a later stage.

**Tail values mirror Tika's `String[]`-backed metadata model**: every entry is a typed array, tagged by the declared `Property` element type, so declared integer sequences (`pdf:charsPerPage`, `tiff:BitsPerSample`) arrive as int64s per element and dates as `Timestamp`s. Untyped keys stay strings, never guessed; if any element refuses its declared type the whole entry falls back to strings. No reflection anywhere — the mapping is an explicit compile-time table (`TikaCoreProperties.TITLE` → `metadata.title`), so reviewing it is reading one file.

**Coverage**: 31 mapper unit tests (tagging contract, DC mapping, builder edge cases, a registered-transformer test demonstrating the stage-2 mechanism), 10 in-process tika-grpc tests including a unary round trip through the real forked pipes worker, and a live-server e2e case asserting typed booleans/integers/strings over the wire.

Deferred to follow-up PRs per the staged plan: per-format transformers, the structured content tree, embedded-document recursion.

### CODE REVIEW

**Comment by krickert** [comment]
@nddipiazza new try - 

Minimal scope, good document shape, no reflection.  Followed your feedback - most of the code is tests.

The mapping is a greppable, compile-time-checked table (TikaCoreProperties.TITLE → metadata.title)

Let's give it another look?

**Comment by tballison** [comment]
From a quick read, this looks great. I'm struggling with how to migrate from what we have to a more structured approach given what I'm hoping is a 4.0.0 release fairly soon.

I personally don't want to hold 4.0.0 for this, but I do think this is important.

Apologies if you've done this in previous iterations, but would it be possible to keep grpc as is and then work on a `tika.v2` service that covers this work and the potential follow on work. We can then release `tika.v2` when this is finalized enough.

Is that too complicated?

@krickert  and @nddipiazza wdyt?

**Comment by krickert** [comment]
> From a quick read, this looks great. I'm struggling with how to migrate from what we have to a more structured approach given what I'm hoping is a 4.0.0 release fairly soon.
> 
> I personally don't want to hold 4.0.0 for this, but I do think this is important.
> 
> Apologies if you've done this in previous iterations, but would it be possible to keep grpc as is and then work on a `tika.v2` service that covers this work and the potential follow on work. We can then release `tika.v2` when this is finalized enough.
> 
> Is that too complicated?
> 
> @krickert and @nddipiazza wdyt?

The v2 idea is spot on. The strange lesson I had by transitioning from a java kingdom to a serialized schema IDL is that the models should never break.   I used to purposefully break my APIs to force customers to write the data how I wanted them to.  Now I see the light.

Protobufs provide an "out" for most of it, and the ultimate out you get is when you break the contract and just add a "v2" so your "v1" shapes can still serialize. Then it'll be up to you to map it. Otherwise data structures are always additive and you deprecate or retire old fields.

So your suggestion is exactly what the "v{n}" package name is for. We can release a new point version of the tika engine separately from the model version since they play well together.

And I don't mind as many iterations as we need. The API shape is something we get stuck with, and the Tika team is the one that defines the shape after parsing.

There are two schools of data structuring that come out of parsing: data shape (the physical look of it like bounding boxes) and then the data surface (NLP analysis, tokens, the chunks plus body, and governed metadata). I've found in parser-land that it's best to separate these two concerns. Docling falls more into the first category (layout-first, NLP second) while tika seems to follow the second (NLP first, layout second). Both sorta flirt with each other as far as data goes, but that tends to be where each parser aims to be.

There are ways to more accurately combine them, but don't ever try to pull an Icarus and fly too close to a unified single model. If you try, you'll end up just angering both camps to use it. The idea is that you share metadata or link between the two data views.

A lot of theory there, but hope this makes sense.. 

BTW - if you want instant docling-looking data using diskless OCR,  [this is a grpc service I slapped together](https://github.com/ai-pipestream/gRParse.git) that's faster than docling (nvidia/IBM gfx compatible) but returns the same data and [this is a grpc service](https://github.com/ai-pipestream/grPOIc) that wraps poi into grpc.

**Comment by nddipiazza** [comment]
@krickert i agree a v2 is needed. we had people using v1 so that was going to be needed regardless because actually tika v3 had a grpc contract that a few groups are using which is compatible with the contract we have in 4.x

**Comment by krickert** [comment]
> @krickert i agree a v2 is needed. we had people using v1 so that was going to be needed regardless because actually tika v3 had a grpc contract that a few groups are using which is compatible with the contract we have in 4.x

I'll go ahead and make the protos in a new package. I'd encourage you to Mark the old ones as deprecated but I'll leave that up to you.  

Are you okay if that's what I do with this pass or is there any additional things you'd like me to change before I start.

**Comment by dpol1** [comment]
rebased on today's main: squashed the commits and fixed the conflicts, plus the API moves (`HttpHeaders` split, `tk:` minting, the `ClimateForecast` rename, `CreativeCommons` removal). one real fix rode along: the digest key had drifted (`SHA256` vs the javaName `SHA-256`) so `origin.sha256` was always empty - now derived from `DigestDef`.

**Comment by dpol1** [comment]
Rebased on `main` 1e3d8f88. One thing changed inside the first commit: #3108's `parse_context_json` gate (`buildRequestParseContext`) got a package-private overload so the v2 RPCs go through it too, and `runFetchAndParse` takes the validated `ParseContext` instead of the two JSON strings. Commits 2 and 3 are byte-identical.

File: tika-grpc-mapper/src/main/java/org/apache/tika/grpc/mapper/DocumentBuilder.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        if (pipesStatus == null || pipesStatus.isEmpty()) {
+            return ParseStatus.Status.UNSPECIFIED;
+        }
+        return switch (pipesStatus) {
+            case "EMPTY_OUTPUT", "PARSE_SUCCESS", "EMIT_SUCCESS", "EMIT_SUCCESS_PASSBACK" ->
+                    ParseStatus.Status.SUCCESS;
+            case "PARSE_SUCCESS_WITH_EXCEPTION", "PARSE_EXCEPTION_NO_EMIT", "EMIT_SUCCESS_PARSE_EXCEPTION" ->
+                    ParseStatus.Status.PARTIAL;
+            default -> ParseStatus.Status.FAILED;
+        };
```
[the remark this answers is not in the document]
Comment (dpol1) [reply tika-grpc-mapper/src/main/java/org/apache/tika/grpc/mapper/DocumentBuilder.java:168]: Fixed in e690d9f2ed. PARTIAL_TIMEOUT is output cut by the deadline and still emitted (success category in PipesResult), so it maps to PARTIAL; TIMEOUT stays FAILED. Test added to DocumentBuilderTest.

File: tika-grpc/src/main/java/org/apache/tika/pipes/grpc/TikaGrpcV2ServerImpl.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            return;
+        }
+        ParseContext parseContext = v1.buildRequestParseContext(request.getFetcherId(),
+                request.getAdditionalFetchConfigJson(), request.getParseContextJson(),
+                responseObserver);
+        if (parseContext == null) {
+            return;
+        }
+        fetchAndParseImpl(request, parseContext, responseObserver);
+    }
```
[the remark this answers is not in the document]
Comment (dpol1) [reply tika-grpc/src/main/java/org/apache/tika/pipes/grpc/TikaGrpcV2ServerImpl.java:76]: Fixed in e690d9f2ed: onCompleted() after the reply, as the v1 method does. Test added: one reply, no error, exactly one completion.

File: tika-grpc/src/main/java/org/apache/tika/pipes/grpc/TikaGrpcServerImpl.java [hunk 3]
```diff
@@ -357,65 +380,120 @@ private ParseContext buildRequestParseContext(FetchAndParseRequest request,
...
+                    errorMessage,
+                    fields,
+                    primary,
+                    fetchParseTimeMs);
         } catch (IOException | PipesException e) {
             throw new RuntimeException(e);
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
+            return null;
         }
```
[the remark this answers is not in the document]
Comment (dpol1) [reply tika-grpc/src/main/java/org/apache/tika/pipes/grpc/TikaGrpcServerImpl.java:481]: Left as is here. The v1 code on main does the same (restores the interrupt, sends no reply), so this PR does not change it. ParseBytes on TIKA-4795 answers UNAVAILABLE in that case; doing the same for FetchAndParse touches the streaming paths too and is a follow-up.
